EtherKit/Models/TypedData.swift

```diff
@@ -0,0 +1,402 @@
+//
+//  TypedData.swift
+//  EtherKit
+//
+//  Created by Zac Morris on 2018-06-14.
+//
+import BigInt
+import Foundation
+import Marshal
+
+// Represents the definition of a member variable for a struct type
+public struct MemberVariable: Unmarshaling {
+  var name: String
+  var type: String
+
+  public init(object: MarshaledObject) throws {
+    name = try object.value(for: "name")
+    type = try object.value(for: "type")
+  }
+
+  // Encodes this member variable
+  public func encodeMember() -> String {
+    return "\(type) \(name)"
+  }
+}
+
+// Represents the collection of member variables defining a struct type
+public struct StructTypeDefinition: ValueType {
+  public var members: [MemberVariable]
+
+  public static func value(from object: Any) throws -> StructTypeDefinition {
+    let memberArray: [MemberVariable] = try [MemberVariable].value(from: object)
+    return StructTypeDefinition(members: memberArray)
+  }
+
+  // Returns the collection of struct types used as a member variable by this struct type
+  public func getSubStructTypes() -> [String] {
+    let standardTypes: Set = ["bool", "uint", "int", "address", "string", "bytes"]
+
+    let subTypes: [String] = members.reduce([String](), { arr, prop in
+      var subType = prop.type
+      if subType.contains("[") {
+        subType = subType.components(separatedBy: "[")[0]
+      }
+      let matches = standardTypes.contains { type in
+        return subType.starts(with: type)
+      }
+      if !matches {
+        if !arr.contains(subType) {
+          var newArr: [String] = arr
+          newArr.append(subType)
+          return newArr
+        }
+      }
+      return arr
+    })
+    return subTypes
+  }
+
+  // Encodes the member variables for this struct type
+  public func encodeMembers() -> String {
+    let encodedMems = members.map { member in
+      return member.encodeMember()
+    }
+    return "(" + encodedMems.joined(separator: ",") + ")"
+  }
+}
+
+// Extends the dictionary representing type name to the struct type definition
+public extension Dictionary where Key == String, Value == StructTypeDefinition {
+  // Returns the typeHash representation for the named struct type
+  public func getTypeHash(for typeName: String) throws -> Data {
+    guard let mainType = self[typeName] else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+
+    var subTypes: [String] = mainType.getSubStructTypes()
+    var seenTypes: [String] = []
```

Comment:
Better as a set?

Reply:
Actually, if you just make `getSubStructTypes` a set, a lot of this logic disappears below.

Reply:
Then you could just do a map for like:
```
let fullTypes = subTypes
  .map {
    guard let subType = self[$0] else {
      throw
    }
    return subType.getSubStructTypes()
  }
  .joined()
```